activerecord/lib/active_record/result.rb

```diff
@@ -127,14 +129,22 @@ def cast_values(type_overrides = {}) # :nodoc:
     def initialize_copy(other)
       @columns      = columns.dup
       @rows         = rows.dup
-      @column_types = column_types.dup
       @hash_rows    = nil
+      @types        = types.dup
+    end
+
+    def column_types
+      if @types.present?
+        Hash[@columns.zip(@types)]
```

Comment:
This method is sometimes called inside a loop:

https://github.com/rails/rails/blob/a3f62cbbe76f7f75c8442d5bab13a8f1bdfd00a7/activerecord/lib/active_record/relation/calculations.rb#L373

The result should be memoised so that it will only be calculated once, as it was before this change.

Reply:
I wonder if it's worth roflscaling this a bit, similar to `hash_rows` below, since this implementation allocates more intermediate objects than before (a two-element array for each column, and an array to hold them all).

Reply:
Since @st0012 was suggesting it earlier, I roflscaled this by assigning `@column_types` in `initialize`